Fixes #10758 - allow execution limiting by permissions

ares opened the pull request:
There are couple of things that has to be solved
* we have to link host of template_invocation somehow
* autocompletion in filters does not work, didn't have time to find why the path is wrong
* if the execution fails we don't display any nice message, in RunHostsJob we just know that sub task failed

stbenjam commented:
> autocompletion in filters does not work, didn't have time to find why the path is wrong

Maybe you'll to override it like Katello does? https://github.com/Katello/katello/blob/master/lib/katello/plugin.rb#L145

ares commented:
I changed the relation so we can use scoped search for host/hostgroup on TemplateInvocation. Also auto completion now works, index route has to be defined as well, we don't need overrides in this plugin.

iNecas commented:
[test]

iNecas commented:
When running migration, I got:

```
undefined method `host_id=' for #<AddHostIdToTemplateInvocation::FakeTemplateInvocation:0x00000009302f60>/home/inecas/Projects/ws/foreman-rex/foreman_remote_execution/db/migrate/20151215114631_add_host_id_to_template_invocation.rb:20:in `block in up'
/home/inecas/Projects/ws/foreman-rex/foreman_remote_execution/db/migrate/20151215114631_add_host_id_to_template_invocation.rb:12:in `each'
/home/inecas/Projects/ws/foreman-rex/foreman_remote_execution/db/migrate/20151215114631_add_host_id_to_template_invocation.rb:12:in `u
```

iNecas commented on db/migrate/20151215114631_add_host_id_to_template_invocation.rb:
```diff
@@ -0,0 +1,29 @@
+  end
+
+  def up
+    # add_column :template_invocations, :host_id, :integer
```
Forgot to uncomment?

iNecas commented:
What about seeding some pre-defined remote execution role, mainly for documenation purposes - it's not trivial to find out what are all the permissions one needs for making the remote execution for some non-admin user to work

iNecas commented:
I have this filter on execute_job_template: `host.name = host-1.foreman.test OR host_group.name = allow`, I got the execution allowed for the `host-1`, but not for the host in `allow` host_group

iNecas commented:
Nm - the last issue was on my side. To wrap the things up, after uncommenting the lines in migration, it works nicely. I would suggest to seed remote execution role with this permissions (this is a minimal set I needed to get it working):

* Job template	view_job_templates	
* Job invocation	view_job_invocations, create_job_invocations	
* Template invocation	execute_template_invocation		
* Host/managed	view_hosts, edit_hosts	
* Smart proxy	view_smart_proxies

iNecas commented:
Tested and works nicely (after some fixes described above), ACK after the comments are addressed

ares commented:
I fixed the migration :bug: :hammer: and added two roles, the edit_hosts permissions would be only useful to make checkboxes on host page active, while user can still assign mupltiple hosts using search in invocation form. I think we shouldn't add this permission to the role. The reason is editing hosts would be too powerful (changing IP to get to different host etc). Otherwise I got the same permissions set.

Please give it another look and let's see what Mr. Jenkins say.

iNecas commented:
ACK, APJ

ares commented:
Rubocop should be happy now

iNecas commented:
[test] - hitting the "HostIntegrationTest::create new host page.test_0003_enables override of hostgroup defaults" starts to be annoying

ares commented:
re[test]

iNecas commented:
The jenkins is pretty false positive right now with the "HostIntegrationTest::create new host page.test_0003_enables override of hostgroup defaults", I'm ok with merging and waiting for it to pass in master

ares commented:
Merging rebased version, thanks for the review.
